x/ccv/consumer/keeper/relay_test.go

```diff
@@ -109,44 +115,87 @@ func (suite *KeeperTestSuite) TestOnRecvVSCPacket() {
 		},
 		{
 			"invalid packet: different destination channel than provider channel",
-			channeltypes.NewPacket(pd.GetBytes(), 1, providertypes.PortID, suite.path.EndpointB.ChannelID, consumertypes.PortID, "InvalidChannel",
+			channeltypes.NewPacket(pd.GetBytes(), 1, providertypes.PortID, providerCCVChannelID, consumertypes.PortID, "InvalidChannel",
 				clienttypes.NewHeight(1, 0), 0),
 			types.ValidatorSetChangePacketData{ValidatorUpdates: []abci.ValidatorUpdate{}},
 			types.ValidatorSetChangePacketData{ValidatorUpdates: []abci.ValidatorUpdate{}},
 			true,
 		},
 	}
 
+	// Instantiate custom keeper with mocks
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+	cdc, storeKey, paramsSubspace, ctx := testkeeper.SetupInMemKeeper(t)
+
+	mockScopedKeeper := testkeeper.NewMockScopedKeeper(ctrl)
+	mockChannelKeeper := testkeeper.NewMockChannelKeeper(ctrl)
+
+	// Setup expected mock calls for final test case
+
+	dummyCap := &capabilitytypes.Capability{}
+	gomock.InOrder(
+
+		mockScopedKeeper.EXPECT().GetCapability(
+			ctx, host.ChannelCapabilityPath(consumertypes.PortID, "InvalidChannel"),
+		).Return(dummyCap, true).Times(1),
+
+		mockChannelKeeper.EXPECT().ChanCloseInit(
+			ctx, consumertypes.PortID, "InvalidChannel", dummyCap,
+		).Return(nil).Times(1),
+	)
+
+	consumerKeeper := testkeeper.GetCustomConsumerKeeperWithMocks(
+		cdc,
+		storeKey,
+		paramsSubspace,
+		mockScopedKeeper,
+		mockChannelKeeper,
+		testkeeper.NewMockPortKeeper(ctrl),
+		testkeeper.NewMockConnectionKeeper(ctrl),
+		testkeeper.NewMockClientKeeper(ctrl),
+		testkeeper.NewMockSlashingKeeper(ctrl),
+		testkeeper.NewMockBankKeeper(ctrl),
+		testkeeper.NewMockAccountKeeper(ctrl),
+		testkeeper.NewMockIBCTransferKeeper(ctrl),
+		testkeeper.NewMockIBCCoreKeeper(ctrl),
+	)
+
+	// Set channel to provider, still in context of consumer chain
+	consumerKeeper.SetProviderChannel(ctx, consumerCCVChannelID)
+	consumerKeeper.SetUnbondingTime(ctx, 100*time.Hour)
+
 	for _, tc := range testCases {
-		ack := suite.consumerChain.App.(*appConsumer.App).ConsumerKeeper.OnRecvVSCPacket(suite.ctx, tc.packet, tc.newChanges)
-		suite.Require().NotNil(ack, "invalid test case: %s did not return ack", tc.name)
+		ack := consumerKeeper.OnRecvVSCPacket(ctx, tc.packet, tc.newChanges)
 
-		if tc.expErrorAck {
-			suite.Require().False(ack.Success(), "invalid test case: %s did not return an Error Acknowledgment", tc.name)
+		if tc.noAck {
+			// if closing channel was successful, no acknowledgement returned, see OnRecvPacketOnUnknownChannel
```

Comment:
This is slightly different than previous expected behavior, an err ack should only be returned if something went wrong with ```ChanCloseInit``` for the ```"invalid packet: different destination channel than provider channel"``` test case

Reply:
@mpoke I don't understand the code this is testing (OnRecvPacketOnUnknownChannel)

Reply:
See `utils.OnRecvPacketOnUnknownChannel`.
```
// OnRecvPacketOnUnknownChannel handles a received packet that was sent
// not on an established CCV channel
```
There was duplicated code on both consumer and provider and I decided to create a helper function.

Reply:
I don't understand why this scenario would ever be exposed to our application logic, given how many layers of abstraction exist in IBC. Why would the IBC module ever give this packet to us? This seems like if you had to write web server code to manually reject all packets with the wrong IP on them instead of that happening in the kernel.

Reply:
So do we assume that for example this condition is never triggered? https://github.com/cosmos/interchain-security/blob/eaa7f6e7ef98fa16b38e6213a194506dc9d50f26/x/ccv/provider/keeper/relay.go#L37

In other words, there is no chance of receiving a packet on an unexpected channel?

Reply:
I'm honestly not sure, but I'd be disappointed if we had to handle stuff like this ourselves.

Reply:
See https://github.com/cosmos/interchain-security/issues/292 for a more detail description of these scenarios.